Sources/PackageCollectionsSigning/Certificate/CertificatePolicy.swift

```diff
@@ -82,19 +86,295 @@ extension CertificatePolicy {
+            let error = CCryptoBoringSSL_X509_verify_cert_error_string(numericCast(CCryptoBoringSSL_X509_STORE_CTX_get_error(x509StoreCtx)))
+            diagnosticsEngine.emit(warning: "The certificate is invalid: \(String(describing: error))")
+            return wrappedCallback(.failure(CertificatePolicyError.invalidCertChain))
+        }
+
+        if certChain.count >= 1, let httpClient = httpClient {
+            // Whether cert chain can be trusted depends on OCSP result
+            ocspClient.checkStatus(certificate: certChain[0], issuer: certChain[1], httpClient: httpClient,
+                                   diagnosticsEngine: diagnosticsEngine, callbackQueue: callbackQueue, callback: callback)
+        } else {
+            wrappedCallback(.success(()))
+        }
     }
+    #endif
 }
 
+#if os(Linux) || os(Windows)
+private let ocspClient = BoringSSLOCSPClient()
+
+private struct BoringSSLOCSPClient {
+    private let resultCache = ThreadSafeKeyValueStore<CacheKey, CacheValue>()
+
+    private let cacheTTL: DispatchTimeInterval
+
+    init(cacheTTL: DispatchTimeInterval = .seconds(300)) {
+        self.cacheTTL = cacheTTL
+    }
+
+    func checkStatus(certificate: Certificate,
+                     issuer: Certificate,
+                     httpClient: HTTPClient,
+                     diagnosticsEngine: DiagnosticsEngine,
+                     callbackQueue: DispatchQueue,
+                     callback: @escaping (Result<Void, Error>) -> Void) {
+        let wrappedCallback: (Result<Void, Error>) -> Void = { result in callbackQueue.async { callback(result) } }
+
+        let ocspURLs = CCryptoBoringSSL_X509_get1_ocsp(certificate.underlying)
+        defer { CCryptoBoringSSL_sk_OPENSSL_STRING_free(ocspURLs) }
+
+        let ocspURLCount = CCryptoBoringSSL_sk_OPENSSL_STRING_num(ocspURLs)
+        // Nothing to do if no OCSP URLs. Use `supportsOCSP` to require OCSP support if needed.
+        guard ocspURLCount > 0 else { return wrappedCallback(.success(())) }
+
+        // Construct the OCSP request
+        let digest = CCryptoBoringSSL_EVP_sha1()
+        guard let certid = OCSP_cert_to_id(digest, certificate.underlying, issuer.underlying),
+            let request = OCSP_REQUEST_new(),
+            OCSP_request_add0_id(request, certid) != nil else {
+            return wrappedCallback(.failure(CertificatePolicyError.ocspSetupFailure))
+        }
+        defer { OCSP_REQUEST_free(request) }
+
+        // Write the request binary to memory bio
+        guard let bio = CCryptoBoringSSL_BIO_new(CCryptoBoringSSL_BIO_s_mem()),
+            i2d_OCSP_REQUEST_bio(bio, request) > 0 else {
+            return wrappedCallback(.failure(CertificatePolicyError.ocspSetupFailure))
+        }
+        defer { CCryptoBoringSSL_BIO_free(bio) }
+
+        // Copy from bio to byte array then convert to Data
+        var count = 0
+        let out = UnsafeMutablePointer<UnsafePointer<UInt8>?>.allocate(capacity: 1)
```

Comment:
This pointer leaks on all code paths. It can be replaced with:

```
var out: UnsafePointer<UInt8>? = nil
```

And then used with `CCryptoBoringSSL_BIO_mem_contents(bio, &out, &count)`.